File: tools/music_audio/qwen3_asr/src/python/qwen3_asr_tool.py
import json
import os
import re
import tempfile
from typing import Optional

from bridges.python.src.sdk.base_tool import BaseTool, ExecuteCommandOptions
from bridges.python.src.sdk.toolkit_config import ToolkitConfig
from tools.music_audio.transcription_schema import TranscriptionOutput, TranscriptionSegment
from bridges.python.src.constants import NVIDIA_LIBS_PATH, PYTORCH_TORCH_PATH

MODEL_NAME = "qwen3-asr-1.7b"
FORCED_ALIGNER_MODEL_NAME = "qwen3-forcedaligner-0.6b"
DEFAULT_SETTINGS = {}
REQUIRED_SETTINGS = []


class Qwen3ASRTool(BaseTool):
    """
    Example output format:

    I noticed the app has a very mobile-first feel.
    [0.08-0.16s] I
    [0.16-0.64s] noticed
    """

    TOOLKIT = "music_audio"

    def __init__(self):
        super().__init__()
        # Load configuration from central toolkits directory
        self.config = ToolkitConfig.load(self.TOOLKIT, self.tool_name)
        self.settings = ToolkitConfig.load_tool_settings(
            self.TOOLKIT, self.tool_name, DEFAULT_SETTINGS
        )
        self.required_settings = REQUIRED_SETTINGS
        self._check_required_settings(self.tool_name)

    @property
    def tool_name(self) -> str:
        # Use the actual config name for toolkit lookup
        return "qwen3_asr"

    @property
    def toolkit(self) -> str:
        return self.TOOLKIT

    @property
    def description(self) -> str:
        return self.config["description"]

    def transcribe_to_file(
        self,
        input_path: str,
        output_path: str,
        device: str = "auto",
        batch_size: int = 4,
        language: str = "auto",
        return_timestamps: bool = True,
        use_forced_aligner: bool = True,
        cuda_runtime_path: Optional[str] = None,
        torch_path: Optional[str] = None,
        chunk_duration: int = 30,
        cpu_batch_size: Optional[int] = None,
    ) -> str:
        """
        Transcribe audio to a file using Qwen3-ASR

        Args:
            input_path: The file path of the audio to be transcribed
            output_path: The desired file path for the transcription output
            device: Device to use for processing (cpu, cuda, auto)
            batch_size: Batch size for processing
            language: Language code for transcription (auto, en, fr, etc.)
            return_timestamps: Whether to return timestamps in output
            use_forced_aligner: Whether to use the forced aligner model
            cuda_runtime_path: Path to CUDA runtime directory (Linux/Windows only)
            torch_path: Path to PyTorch installation directory
            chunk_duration: Chunk duration in seconds for long audio
            cpu_batch_size: CPU batch size for long audio

        Returns:
            The path to the transcription file
        """
        try:
            if not os.path.isfile(input_path):
                raise FileNotFoundError(f"Input audio file does not exist: {input_path}")

            os.makedirs(os.path.dirname(output_path) or ".", exist_ok=True)

            model_path = self.get_resource_path(MODEL_NAME)
            forced_aligner_path = None
            nvidia_libs_path = (
                cuda_runtime_path if cuda_runtime_path is not None else NVIDIA_LIBS_PATH
            )
            torch_libs_path = (
                torch_path if torch_path is not None else PYTORCH_TORCH_PATH
            )

            if return_timestamps and use_forced_aligner:
                forced_aligner_path = self.get_resource_path(FORCED_ALIGNER_MODEL_NAME)

            tasks = [
                {
                    "audio_path": input_path,
                    "output_path": output_path,
                }
            ]

            with tempfile.NamedTemporaryFile(
                mode="w", suffix=".json", delete=False, encoding="utf-8"
            ) as temp_file:
                json_file_path = temp_file.name
                json.dump(tasks, temp_file, indent=2, ensure_ascii=False)

            args = [
                "--function",
                "transcribe_audio",
                "--json_file",
                json_file_path,
                "--model_path",
                model_path,
                "--device",
                device,
                "--batch_size",
                str(batch_size),
                "--language",
                language,
                "--return_timestamps",
                "true" if return_timestamps else "false",
                "--chunk_duration",
                str(chunk_duration),
            ]

            if nvidia_libs_path:
                args.extend(["--cuda_runtime_path", nvidia_libs_path])

            if torch_libs_path:
                args.extend(["--torch_path", torch_libs_path])

            if forced_aligner_path:
                args.extend(["--forced_aligner_model_path", forced_aligner_path])

            if cpu_batch_size is not None:
                args.extend(["--cpu_batch_size", str(cpu_batch_size)])

            self.execute_command(
                ExecuteCommandOptions(
                    binary_name="qwen3_asr", args=args, options={"sync": True}
                )
            )

            with open(output_path, "r", encoding="utf-8") as f:
                transcription_content = f.read()

            parsed_output = self.parse_transcription(transcription_content)

            with open(output_path, "w", encoding="utf-8") as f:
                json.dump(parsed_output, f, indent=2, ensure_ascii=False)

            return output_path
        except Exception as e:
            raise Exception(f"Audio transcription failed: {str(e)}")

    def parse_transcription(self, raw_output: str) -> TranscriptionOutput:
        lines = [line.strip() for line in raw_output.split("\n") if line.strip()]

        segments: list[TranscriptionSegment] = []
        segment_regex = re.compile(r"^\[(\d+(?:\.\d+)?)-(\d+(?:\.\d+)?)s\]\s+(.+)$")
        duration = 0.0

        for line in lines:
            match = segment_regex.match(line)
            if match:
                start = float(match.group(1))
                end = float(match.group(2))
                text = match.group(3)

                segments.append(
                    {"from": start, "to": end, "text": text.strip(), "speaker": None}
                )

                if end > duration:
                    duration = end

        if not segments and lines:
            segments.append({"from": 0.0, "to": 0.0, "text": lines[0], "speaker": None})

        return {
            "duration": duration,
            "speakers": [],
            "speaker_count": 0,
            "segments": segments,
            "metadata": {"tool": self.tool_name},
        }

File: tools/music_audio/qwen3_asr/src/nodejs/qwen3_asr-tool.ts
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'

import type { TranscriptionOutput } from '@tools/music_audio/transcription-schema'
import { Tool } from '@sdk/base-tool'
import { ToolkitConfig } from '@sdk/toolkit-config'
import { NVIDIA_LIBS_PATH, PYTORCH_TORCH_PATH } from '@bridge/constants'

type Qwen3ASRTranscriptionOutput = string

const MODEL_NAME = 'qwen3-asr-1.7b'
const FORCED_ALIGNER_MODEL_NAME = 'qwen3-forcedaligner-0.6b'
const DEFAULT_SETTINGS: Record<string, unknown> = {}
const REQUIRED_SETTINGS: string[] = []

interface Qwen3ASRTask {
  audio_path: string
  output_path?: string
}

export default class Qwen3ASRTool extends Tool {
  private static readonly TOOLKIT = 'music_audio'
  private readonly config: ReturnType<typeof ToolkitConfig.load>

  constructor() {
    super()
    // Load configuration from central toolkits directory
    this.config = ToolkitConfig.load(Qwen3ASRTool.TOOLKIT, this.toolName)
    const toolSettings = ToolkitConfig.loadToolSettings(
      Qwen3ASRTool.TOOLKIT,
      this.toolName,
      DEFAULT_SETTINGS
    )
    this.settings = toolSettings
    this.requiredSettings = REQUIRED_SETTINGS
    this.checkRequiredSettings(this.toolName)
  }

  get toolName(): string {
    // Use the actual config name for toolkit lookup
    return 'qwen3_asr'
  }

  get toolkit(): string {
    return Qwen3ASRTool.TOOLKIT
  }

  get description(): string {
    return this.config['description']
  }

  /**
   * Transcribe audio to a file using Qwen3-ASR
   * @param inputPath The file path of the audio to be transcribed
   * @param outputPath The desired file path for the transcription output
   * @param device Device to use for processing (cpu, cuda, auto)
   * @param batchSize Batch size for processing
   * @param language Language code for transcription (auto, en, fr, etc.)
   * @param returnTimestamps Whether to return timestamps in output
   * @param useForcedAligner Whether to use the forced aligner model
   * @param cudaRuntimePath Path to CUDA runtime directory (Linux/Windows only)
   * @param torchPath Path to PyTorch installation directory
   * @param chunkDuration Chunk duration in seconds for long audio
   * @param cpuBatchSize CPU batch size for long audio
   * @returns A promise that resolves with the path to the transcription file
   */
  async transcribeToFile(
    inputPath: string,
    outputPath: string,
    device = 'auto',
    batchSize = 4,
    language = 'auto',
    returnTimestamps = true,
    useForcedAligner = true,
    cudaRuntimePath?: string,
    torchPath?: string,
    chunkDuration = 30,
    cpuBatchSize?: number
  ): Promise<string> {
    let tempDir: string | null = null
    let jsonFilePath: string | null = null

    try {
      const inputStats = await fs.promises.stat(inputPath).catch(() => null)
      if (!inputStats?.isFile()) {
        throw new Error(`Input audio file does not exist: ${inputPath}`)
      }

      await fs.promises.mkdir(path.dirname(outputPath), { recursive: true })

      const modelPath = await this.getResourcePath(MODEL_NAME)
      const forcedAlignerPath =
        returnTimestamps && useForcedAligner
          ? await this.getResourcePath(FORCED_ALIGNER_MODEL_NAME)
          : undefined
      const nvidiaLibsPath = cudaRuntimePath ?? NVIDIA_LIBS_PATH
      const torchLibsPath = torchPath ?? PYTORCH_TORCH_PATH

      const tasks: Qwen3ASRTask[] = [
        {
          audio_path: inputPath,
          output_path: outputPath
        }
      ]

      tempDir = await fs.promises.mkdtemp(
        path.join(os.tmpdir(), 'qwen3_asr_tasks_')
      )
      jsonFilePath = path.join(tempDir, 'tasks.json')

      await fs.promises.writeFile(
        jsonFilePath,
        JSON.stringify(tasks, null, 2),
        'utf8'
      )

      const args = [
        '--function',
        'transcribe_audio',
        '--json_file',
        jsonFilePath,
        '--model_path',
        modelPath,
        '--device',
        device,
        '--batch_size',
        batchSize.toString(),
        '--language',
        language,
        '--return_timestamps',
        returnTimestamps ? 'true' : 'false',
        '--chunk_duration',
        chunkDuration.toString()
      ]

      if (nvidiaLibsPath) {
        args.push('--cuda_runtime_path', nvidiaLibsPath)
      }

      if (torchLibsPath) {
        args.push('--torch_path', torchLibsPath)
      }

      if (forcedAlignerPath) {
        args.push('--forced_aligner_model_path', forcedAlignerPath)
      }

      if (cpuBatchSize) {
        args.push('--cpu_batch_size', cpuBatchSize.toString())
      }

      await this.executeCommand({
        binaryName: 'qwen3_asr',
        args,
        options: { sync: true }
      })

      const transcriptionContent = await fs.promises.readFile(
        outputPath,
        'utf-8'
      )
      const parsedOutput = this.parseTranscription(transcriptionContent)

      await fs.promises.writeFile(
        outputPath,
        JSON.stringify(parsedOutput, null, 2),
        'utf8'
      )

      return outputPath
    } catch (error: unknown) {
      throw new Error(`Audio transcription failed: ${(error as Error).message}`)
    }
  }

  private parseTranscription(
    rawOutput: Qwen3ASRTranscriptionOutput
  ): TranscriptionOutput {
    const lines = rawOutput
      .split('\n')
      .map((line) => line.trim())
      .filter((line) => line.length > 0)

    const segments: TranscriptionOutput['segments'] = []
    const segmentRegex = /^\[(\d+(?:\.\d+)?)-(\d+(?:\.\d+)?)s\]\s+(.+)$/
    let duration = 0

    for (const line of lines) {
      const match = line.match(segmentRegex)
      if (match && match[1] && match[2] && match[3]) {
        const start = parseFloat(match[1])
        const end = parseFloat(match[2])

        segments.push({
          from: start,
          to: end,
          text: match[3].trim(),
          speaker: null
        })

        if (end > duration) {
          duration = end
        }
      }
    }

    if (segments.length === 0 && lines.length > 0) {
      segments.push({
        from: 0,
        to: 0,
        text: lines[0] ?? '',
        speaker: null
      })
    }

    return {
      duration,
      speakers: [],
      speaker_count: 0,
      segments,
      metadata: {
        tool: this.toolName
      }
    }
  }
}
